### PR TITLE
add SNLI model

### PR DESCRIPTION
**Pull request by xuyige**
there are serval changes in this pull request,

1, clean up some code

2, modify some modules, including lstm, linear
for example, the forward function of encoder.lstm was changed.
before:
```
    def forward(self, x):
        x, _ = self.lstm(x)
        return x
```
now:
```
    def forward(self, x, h0=None, c0=None):
        if h0 is not None and c0 is not None:
            x, (ht, ct) = self.lstm(x, (h0, c0))
        else:
            x, (ht, ct) = self.lstm(x)
        if self.get_hidden:
            return x, (ht, ct)
        else:
            return x
```
after modifying this, the former usage is also compatible.

3. fix some mislead comments in embed loader and rename some variables

4. add SNLI model

5. modify the index_field function in Vocabulary
before:
```
    def index_field(self, field_name, vocab):
        for ins in self:
            ins.index_field(field_name, vocab)
        return self
```
now:
```
    def index_field(self, field_name, vocab):
        if isinstance(field_name, str) and isinstance(vocab, Vocabulary):
            field_list = [field_name]
            vocab_list = [vocab]
        else:
            classes = (list, tuple)
            assert isinstance(field_name, classes) and isinstance(vocab, classes) and len(field_name) == len(vocab)
            field_list = field_name
            vocab_list = vocab
         for name, vocabs in zip(field_list, vocab_list):
            for ins in self:
                ins.index_field(name, vocabs)
        return self
```
if we need to index serval fields, we can pass two list to this function instead of calling this function serval times. meanwhile, the former usage that passing both a string and a Vocabulary is also compatible.

### CODE REVIEW

**Comment by FengZiYjun** [comment]
Many thanks 😃 
For LSTM, the change is reasonable.
For index method, "pass two list to this function instead of calling this function serval times". Is this an expected behavior? @choosewhatulike

**Comment by choosewhatulike** [comment]
> For index method, "pass two list to this function instead of calling this function serval times". Is this an expected behavior? @choosewhatulike

it's good to add new features. and it support old APIs

File: fastNLP/models/snli.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        # v_mlp = F.tanh(self.mlp_layer1(v))  # v_mlp: [B, H]
+        # prediction = self.mlp_layer2(v_mlp)  # prediction: [B, N]
+
+        prediction = F.tanh(self.output(v))  # prediction: [B, N]
+
+        return prediction
+
+    @staticmethod
+    def calc_bi_attention(in_x1, in_x2, x1_len, x2_len):
```
Comment (FengZiYjun) [inline fastNLP/models/snli.py]: this static function can be extracted as a module

File: fastNLP/models/snli.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        attention_b = F.softmax(attention_matrix + b_mask, dim=1)  # [batch_size, x1_seq_len, x2_seq_len]
+
+        out_x1 = torch.bmm(attention_a, in_x2)  # [batch_size, x1_seq_len, hidden_size]
+        attention_b_t = torch.transpose(attention_b, 1, 2)
+        out_x2 = torch.bmm(attention_b_t, in_x1)  # [batch_size, x2_seq_len, hidden_size]
+
+        return out_x1, out_x2
+
+    @staticmethod
+    def mean_pooling(tensor, mask, dim=0):
```
Comment (FengZiYjun) [inline fastNLP/models/snli.py]: these two static methods can also be extracted as modules

Comment (xuyige) [reply fastNLP/models/snli.py]: yes, @xpqiu had pointed out this problem, these method will be moved into moudules in the next version

**Comment by xuyige** [comment]
> Many thanks 😃
> For LSTM, the change is reasonable.
> For index method, "pass two list to this function instead of calling this function serval times". Is this an expected behavior? @choosewhatulike
> 
> Everything goes fine. Ready to merge.

if we need to index several fields at the same time, we can call the index function once, which may reduce the lines of code and seems more clear.

**Pull request merged**